### PR TITLE
Add mertrics

### PR DESCRIPTION
**Pull request by bartlomieju**
This PR is WIP for implementation of performance metrics.

Any help or feedback welcome.

EDIT: Ref #883

### CODE REVIEW

**Comment by bartlomieju** [comment]
I hit the wall trying to update values of `IsolateMetrics`. I think `IsolateState` has to be made mutable inside `pre_dispatch`, but it's beyond my Rust knowledge ATM. Any tips @ry ?

File: js/metrics.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+import * as msg from "gen/msg_generated";
+import { flatbuffers } from "flatbuffers";
+import { assert } from "./util";
+import * as dispatch from "./dispatch";
+
+interface DenoMetrics {
```
Comment (ry) [inline js/metrics.ts]: Don't prefix with "Deno", just "Metrics".

File: src/isolate.rs [hunk 2]
```diff
@@ -69,6 +70,40 @@ impl IsolateState {
   }
 }
 
+pub struct IsolateMetrics {
```
Comment (ry) [inline src/isolate.rs]: s/IsolateMetrics/Metrics/g

File: src/ops.rs [hunk 3]
```diff
@@ -1148,3 +1149,34 @@ fn op_dial(
...
+  ok_future(serialize_response(
+    cmd_id,
+    builder,
+    msg::BaseArgs {
+      inner: Some(inner.as_union_value()),
+      inner_type: msg::Any::MetricsRes,
+      ..Default::default()
+    },
+  ))
+}
```
Comment (ry) [inline src/ops.rs]: Looks good.

File: src/isolate.rs [hunk 4]
```diff
@@ -93,6 +128,7 @@ impl Isolate {
         argv: argv_rest,
         flags,
         tx: Mutex::new(Some(tx)),
+        metrics: Metrics::new(),
```
Comment (hayd) [inline src/isolate.rs]: Does this need to be a Mutex?

**Comment by bartlomieju** [comment]
I've tried to implement first metric (counting ops), but I've hit this:
```
 --> ../../src/isolate.rs:307:5
    |
307 |     state.metrics.increment_ops_executed();
    |     ^^^^^ cannot move out of borrowed content
```

File: src/isolate.rs [hunk 5]
```diff
@@ -69,6 +70,40 @@ impl IsolateState {
...
+  pub fn new() -> Metrics {
+    Metrics {
+      ops_executed: 0,
+      bytes_recv: 0,
+      bytes_sent: 0
+    }
+  }
+
+  #[allow(dead_code)]
+  fn increment_ops_executed(mut self) {
```
Comment (bluetech) [inline src/isolate.rs]: You likely want `&mut self` here and below. Plain `mut self` moves `self` into the function (`self` is "consumed"), while `&mut self` takes `self` by reference.

File: src/isolate.rs [hunk 6]
```diff
@@ -259,11 +295,18 @@ extern "C" fn pre_dispatch(
 
   if is_sync {
     // Execute op synchronously.
+    let state = isolate.state.clone();
```
Comment (bluetech) [inline src/isolate.rs]: This is a style preference, but I prefer `Arc::clone(&isolate.state)` over `isolate.state.clone()`. The two are exactly equivalent, however the first makes it clear that it's an Arc that's cloned and not the data itself.

File: src/isolate.rs (hunk 4 above)
Comment (hayd) [reply src/isolate.rs]: To clarify, this comment was about the metrics line, right?

File: src/isolate.rs (hunk 5 above)
Comment (bartlomieju) [reply src/isolate.rs]: Thanks, however when I do that I get 
```
error[E0596]: cannot borrow field of immutable binding as mutable
   --> ../../src/isolate.rs:307:5
    |
307 |     state.metrics.increment_ops_executed();
    |     ^^^^^^^^^^^^^ cannot mutably borrow field of immutable binding
```

Not sure how can I make `metrics` a mutable binding of `state` since it's field in struct

File: src/isolate.rs (hunk 6 above)
Comment (bartlomieju) [reply src/isolate.rs]: Agreed

File: src/isolate.rs (hunk 5 above)
Comment (bluetech) [reply src/isolate.rs]: I believe this will be fixed if you wrap the `metrics` field in a `Mutex` as suggested above.

The problem is that you have an `Arc` which has shared ownership, while mutating requires unique ownership. If you *know* that you are the only one holding the `Arc`, you can use [`Arc::get_mut`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.get_mut), but normally you don't know that so you need a `Mutex`.

Comment (bartlomieju) [reply src/isolate.rs]: Thanks @bluetech! With your advice I managed to get it working!

File: src/isolate.rs [hunk 7]
```diff
@@ -56,6 +56,7 @@ pub struct IsolateState {
   pub argv: Vec<String>,
   pub flags: flags::DenoFlags,
   tx: Mutex<Option<mpsc::Sender<(i32, Buf)>>>,
+  pub metrics: Mutex<Option<Metrics>>,
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply src/isolate.rs]: TBH I took example of `tx` - good point I'll fix that

File: src/isolate.rs [hunk 8]
```diff
@@ -67,6 +68,51 @@ impl IsolateState {
...
+    Metrics {
+      ops_executed: 0,
+      bytes_recv: 0,
+      bytes_sent: 0
+    }
+  }
+
+  #[allow(dead_code)]
+  fn increment_ops_executed(&mut self) {
+    assert!(self.ops_executed >= 0);
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply src/isolate.rs]: Again I do agree, my first iteration used unsigned values but to get it working I used `i64` but got errors down the road. Will do that as well

File: src/isolate.rs [hunk 9]
```diff
@@ -253,20 +300,26 @@ extern "C" fn pre_dispatch(
     )
   };
 
+  let bytes_recv = data_buf.data_len as i32;
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply src/isolate.rs]: Probably not, will rework for unsigned type as well.

**Comment by ry** [comment]
Looking good - please some unit tests: `js/metrics_test.ts`

**Comment by bartlomieju** [comment]
@bluetech thanks for review, I learned a lot! @ry I added simple test case for metrics, tell me if I should add more of them

**Comment by ry** [comment]
@bartlomieju that’s good for the JS side.

I’d like to see another test in src/isolate.rs
There should be a test case in there you can copy

**Comment by bartlomieju** [comment]
@ry I've added the test case in Rust

File: src/isolate.rs [hunk 10]
```diff
@@ -386,4 +387,38 @@ mod tests {
...
+          const m = new Uint8Array([4, 5, 6]);
+          libdeno.send(m, m);
+        "#,
+        ).expect("execute error");
+      isolate.event_loop();
+      let g = isolate.state.metrics.lock().unwrap();
+      let metrics = g.deref();
+      assert_eq!(metrics.ops_executed, 1);
+      assert_eq!(metrics.bytes_sent, 3);
+      assert_eq!(metrics.bytes_recv, 3);
```
Comment (ry) [inline src/isolate.rs]: Very nice.

File: src/isolate.rs [hunk 11]
```diff
@@ -386,4 +387,38 @@ mod tests {
     let op = Box::new(futures::future::ok(control));
     (true, op)
   }
+
+  #[test]
+  fn test_metrics() {
+    let argv = vec![String::from("./deno"), String::from("hello.js")];
+    let mut isolate = Isolate::new(argv, metrics_dispatch);
+    tokio_util::init(|| {
+      isolate
```
Comment (ry) [inline src/isolate.rs]: Please check the metrics here too

File: src/isolate.rs [hunk 12]
```diff
@@ -346,4 +387,47 @@ mod tests {
...
+    });
+  }
+
+  fn metrics_dispatch(
+    _isolate: &mut Isolate,
+    _control: &[u8],
+    _data: &'static mut [u8],
+  ) -> (bool, Box<Op>) {
+    // Send back some sync response
+    let vec: Vec<u8> = vec![1, 2, 3];
```
Comment (ry) [inline src/isolate.rs]: Send back 4 bytes instead of 3, so we check that recv vs send is different.

File: src/isolate.rs [hunk 13]
```diff
@@ -346,4 +387,47 @@ mod tests {
...
+        assert_eq!(metrics.bytes_sent, 0);
+        assert_eq!(metrics.bytes_recv, 0);
+      }
+
+      isolate
+        .execute(
+          "y.js",
+          r#"
+          const m = new Uint8Array([4, 5, 6]);
+          libdeno.send(m, m);
```
Comment (ry) [inline src/isolate.rs]: Send a different Uint8Array for the second argument of libdeno.send...

```
const control = new Uint8Array([4, 5, 6]);
const data = new Uint8Array([42, 43, 44, 45, 46]);
libdeno.send(control, data);
```

**Comment by bartlomieju** [comment]
@ry I updated the tests, however I don't know why AppVeyor is failing :(

**Comment by kevinkassimo** [comment]
@bartlomieju cuz there are merge conflicts against master

**Comment by bartlomieju** [comment]
I've rebased branch against master

File: src/ops.rs [hunk 14]
```diff
@@ -25,6 +25,7 @@ use std;
 use std::fs;
 use std::net::{Shutdown, SocketAddr};
 #[cfg(any(unix))]
+use std::ops::Deref;
 use std::os::unix::fs::PermissionsExt;
```
Comment (kevinkassimo) [inline src/ops.rs]: Move `#[cfg(any(unix))]` above this line, to avoid windows trying to import unix only extensions

Comment (bartlomieju) [reply src/ops.rs]: My bad, sorry!

File: src/isolate.rs [hunk 15]
```diff
@@ -330,4 +372,49 @@ mod tests {
...
+          libdeno.send(control, data);
+        "#,
+        )
+        .expect("execute error");
+      isolate.event_loop();
+      let g = isolate.state.metrics.lock().unwrap();
+      let metrics = g.deref();
+      assert_eq!(metrics.ops_executed, 1);
+      assert_eq!(metrics.bytes_recv, 5);
+      assert_eq!(metrics.bytes_sent, 4);
```
Comment (ry) [inline src/isolate.rs]: Reverse your usage of "sent" and "recv" - it should be from the perspective of the javascript user.

File: src/isolate.rs [hunk 16]
```diff
@@ -330,4 +372,49 @@ mod tests {
     let op = Box::new(futures::future::ok(control));
     (true, op)
   }
+
+  #[test]
+  fn test_metrics() {
```
Comment (ry) [inline src/isolate.rs]: Rename to `test_metrics_sync`

File: js/metrics_test.ts [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+
+test(function metrics() {
+  const metrics1 = deno.metrics();
+  assert(metrics1.opsExecuted > 0);
+  assert(metrics1.bytesRecv > 0);
+  assert(metrics1.bytesSent > 0);
+
+  const metrics2 = deno.metrics();
+  assert(metrics2.opsExecuted > metrics1.opsExecuted);
+  assert(metrics2.bytesSent > metrics1.bytesRecv);
```
Comment (bluetech) [inline js/metrics_test.ts]: Sent -> Recv I assume.

File: src/isolate.rs [hunk 18]
```diff
@@ -66,6 +68,35 @@ impl IsolateState {
     let tx = maybe_tx.unwrap();
     tx.send((req_id, buf)).expect("tx.send error");
   }
+
+  fn update_metrics(&self, bytes_recv: u64, bytes_sent: u64) {
+    let mut g = self.metrics.lock().unwrap();
+    let metrics = g.deref_mut();
```
Comment (bluetech) [inline src/isolate.rs]: You don't normally need to call `deref_mut`, since Rust automatically inserts it. Does it work if you do the following?

```rust
let mut metrics = self.metrics.lock().unwrap();
metrics.update(bytes_recv, bytes_sent);
```

File: src/isolate.rs [hunk 19]
```diff
@@ -66,6 +68,35 @@ impl IsolateState {
...
+
+    metrics.update(bytes_recv, bytes_sent);
+  }
+}
+
+pub struct Metrics {
```
Comment (bluetech) [inline src/isolate.rs]: You might consider adding `[derive(Default)]` and then using `Default::default()` in `new()` (or even omit `new()` completely). I'm not sure if it's better.

File: src/isolate.rs [hunk 20]
```diff
@@ -236,29 +268,37 @@ extern "C" fn pre_dispatch(
...
     // Execute op synchronously.
     let buf = tokio_util::block_on(op).unwrap();
-    if buf.len() != 0 {
+    let buf_size = buf.len();
+    if buf_size != 0 {
       // Set the synchronous response, the value returned from isolate.send().
       isolate.respond(req_id, buf);
     }
+
+    let state = Arc::clone(&isolate.state);
```
Comment (bluetech) [inline src/isolate.rs]: Are you sure you need the `Arc::clone` here? From the looks of it you should be able to just do `isolate.state.update_metrics(...)` in the synchronous case.

File: src/isolate.rs [hunk 21]
```diff
@@ -330,4 +372,49 @@ mod tests {
...
+
+  #[test]
+  fn test_metrics() {
+    let argv = vec![String::from("./deno"), String::from("hello.js")];
+    let mut isolate = Isolate::new(argv, metrics_dispatch);
+    tokio_util::init(|| {
+      // verify that metrics have been properly initialized
+      {
+        let g = isolate.state.metrics.lock().unwrap();
+        let metrics = g.deref();
```
Comment (bluetech) [inline src/isolate.rs]: Similar comment here regarding `Deref`, normally the implicit deref coercion takes care of things.

File: src/isolate.rs [hunk 22]
```diff
@@ -330,4 +372,49 @@ mod tests {
...
+          const data = new Uint8Array([42, 43, 44, 45, 46]);
+          libdeno.send(control, data);
+        "#,
+        )
+        .expect("execute error");
+      isolate.event_loop();
+      let g = isolate.state.metrics.lock().unwrap();
+      let metrics = g.deref();
+      assert_eq!(metrics.ops_executed, 1);
+      assert_eq!(metrics.bytes_recv, 5);
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply src/isolate.rs]: Thanks for explanation, wasn't sure if I should include control data in the counter. Separate ones makes most sense.

File: src/isolate.rs (hunk 18 above)
Comment (bartlomieju) [reply src/isolate.rs]: Thanks, worked like a charm!

File: src/isolate.rs (hunk 20 above)
Comment (bartlomieju) [reply src/isolate.rs]: Yes, thanks

File: src/isolate.rs (hunk 19 above)
Comment (bartlomieju) [reply src/isolate.rs]: I used that, thanks!

File: js/metrics_test.ts [hunk 23]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+import { test, assert } from "./test_util.ts";
+import * as deno from "deno";
+
+test(function metrics() {
+  // TODO: metrics command does not send any data buffer to privileged side
+  // should we call other method that does to present it here?
```
Comment (bartlomieju) [inline js/metrics_test.ts]: @ry what do you think about that? Should we just leave it as is now?

Comment (ry) [reply js/metrics_test.ts]: Sorry, I don't understand the TODO. Can you rephrase?

File: src/isolate.rs [hunk 24]
```diff
@@ -221,6 +256,10 @@ extern "C" fn pre_dispatch(
   control_buf: libdeno::deno_buf,
   data_buf: libdeno::deno_buf,
 ) {
+  // for metrics
+  let control_bytes_sent = control_buf.data_len as u64;
+  let data_bytes_sent = data_buf.data_len as u64;
```
Comment (ry) [inline src/isolate.rs]: It would be good to update the control_bytes_sent and data_bytes_sent before the op is dispatched.
Async ops might take a long time to complete.
It would be good to have a test for this behavior too.

File: js/metrics_test.ts (hunk 23 above)
Comment (bartlomieju) [reply js/metrics_test.ts]: Bingo

File: src/isolate.rs (hunk 24 above)
Comment (bartlomieju) [reply src/isolate.rs]: I've added the logic for keeping track of them. Not sure how to test it with async operation though.

**Comment by ry** [comment]
@bartlomieju it's a race condition. continuing in #963
